ledger-core/vanilla/cachekit/core.go

```diff
@@ -0,0 +1,376 @@
+
+	if p.recent == nil {
+		p.recent = &generation{
+			start:  age,
+			end:    age,
+		}
+		p.recent.init(p.strat.NextGenerationCapacity(-1, -1))
+
+		p.rarest = p.recent
+		p.oldest = p.recent
+		return p.recent.Add(index, age)
+	}
+
+	p.checkGen(age, added)
+	return p.recent.Add(index, age)
+}
+
+func (p *Core) checkGen(age Age, added bool) {
+	trimmed := false
+	if p.trimEach && added {
+		trimCount := p.strat.CanTrimEntries(p.alloc.Count(), // is already allocated
+			p.recent.end, p.oldest.start)
+
+		p.trimEntries(trimCount)
+		trimmed = trimCount > 0
+	}
+
+	if p.hitLimit > 0 {
+		p.hitLimit--
+	}
+	curLen, curCap := len(p.recent.access), cap(p.recent.access)
+
+	switch {
+	case curLen == curCap:
+		p.hitLimit, p.ageLimit = 0, 0 // enforce call to CanAdvanceGeneration on next update
+	case p.hitLimit == 0 || p.ageLimit <= age:
+		createGen := false
+		createGen, p.hitLimit, p.ageLimit = p.strat.CanAdvanceGeneration(curLen, curCap, p.hitLimit)
+		if createGen {
+			break
+		}
+		fallthrough
+	default:
+		return
+	}
+
+	newGen := &generation{
+		genNo: p.recent.genNo + 1,
+		start: age,
+	}
+	newGen.init(p.strat.NextGenerationCapacity(curLen, curCap))
+
+	if newGen.start < p.recent.end {
+		newGen.start = p.recent.end
+	}
+	newGen.end = newGen.start
+
+	p.recent.next = newGen
+	p.recent = newGen
+
+	if !trimmed {
+		totalCount := p.alloc.Count()
+
+		trimCount := p.strat.CanTrimEntries(totalCount, p.recent.end, p.oldest.start)
+		p.trimEntries(trimCount)
+
+		if trimCount <= 0 {
+			trimCount = p.strat.CanTrimGenerations(totalCount, p.recent.genNo - p.rarest.genNo + 1,
+				p.recent.end, p.rarest.start, p.oldest.start)
+
+			p.trimGenerations(trimCount)
+		}
+	}
+}
+
+func (p *Core) trimEntries(count int) {
+	for count > 0 && p.oldest != p.rarest {
+		count = p.oldest.trimOldest(p, count)
+		if len(p.oldest.access) == 0 {
+			p.oldest = p.oldest.next
+		}
+	}
+
+	for count > 0 && p.recent != p.rarest {
+		count = p.rarest.trimRarest(p, count)
+		if len(p.rarest.access) == 0 {
+			p.rarest = p.rarest.next
+			p.oldest = p.rarest
+		}
+	}
+}
+
+func (p *Core) trimGenerations(count int) {
+	prev := p.oldest
+	if prev == p.rarest || prev.next != p.rarest {
+		prev = nil
+	}
+
+	for ;count > 0 && p.recent != p.rarest; count-- {
+		p.rarest.trimGeneration(p, prev)
+
+		if len(p.rarest.access) == 0 {
+			p.rarest = p.rarest.next
+			if prev != nil {
+				prev.next = p.rarest
+			}
+		} else {
+			prev = p.rarest
+			p.rarest = p.rarest.next
+		}
+	}
+}
+
+func (p *Core) ResetGenerations() {
+
+}
+
+/*******************************************/
+
+type generation struct {
+	access []uint32 // Index
+	fence  map[uint32]struct{}
+	start  Age
+	end    Age
+	next   *generation
+	genNo  GenNo
+}
+
+func (p *generation) init(capacity int, loadFence bool) {
+	p.access = make([]uint32, 0, capacity)
+	if loadFence {
+		p.fence = make(map[uint32]struct{}, capacity)
+	}
+}
+
+func (p *generation) Add(index Index, age Age) GenNo {
```

Comment:
public method on a private strucr

Reply:
Public methods on private structures can have two roles:
1. for interfaces
2. to show that methods should be used outside

only problem I see here is inconsistency: `Add` is 'public', but `trimOldest` is not.

Reply:
what I meant is that `generation` is tightly coupled with `Core`, can not function outside of it and can not be accessed from outside. There are no public methods that return generation object on the Core API.

Reply:
I use public here to indicate methods used by another class/struct (Core in this case) as Go has no way to separate private and protected methods